src/modules/avformat/factory.c

```diff
@@ -229,6 +229,12 @@ static void add_parameters( mlt_properties params, void *object, int req_flags,
 			mlt_properties_set( p, "type", "string" );
 			break;
 		}
+		mlt_properties_set( p, "mutable", "yes" );
+#if LIBAVUTIL_VERSION_INT >= ((56<<16)+(35<<8)+101)
+		if ((opt->flags & AV_OPT_FLAG_RUNTIME_PARAM)) {
```

Comment:
According to this implementation, all animated properties must be numbers. But I do not think this is always true for AV commands. Some AV commands are not represented as numbers. For example, see "width_type" for the equalizer filters. It is a string parameter. 

Maybe there is nothing we can do about that since I do not see a way to query the FFMpeg API to find out if a parameter is a number or string. Maybe we can maintain a blacklist. Or maybe users just need to know that they can not actually animate string parameters.

Reply:
[AVOptions](https://ffmpeg.org/doxygen/trunk/structAVOption.html) do have a type. It is true this limits it to numeric only, but like I said a lot of the AVOptions for a filter have the type as string to support expressions even though the expression must evaluate to numeric to be usable. So, I think this might need to be a sacrifice for the majority. I do not want to maintain our own list of types for each filter's options because it is version-sensitive and has been growing a lot. We need to think less about users and more about developers. I have been saying for a while now that MLT is for developers. Thus, the developers need to be careful about how it is used and avoid using keyframes for actual string parameters. In case a developer tries to expose everything generically, it will be a problem they may need to workaround somehow.

Reply:
> the developers need to be careful about how it is used and avoid using keyframes for actual string parameters

If I understand the implementation correctly, even non-animated string properties will be converted to a double. I expect that "real" string properties that are not numbers will not survive that conversion. So I think it will be impossible to set the "width_type" parameter (whether animated or not).

Maybe we can accept that limitation for now and state it in the documentation. Or maybe we could add a new is_number() function to mlt_properties and use some heuristic to detect that a string is a number.

Reply:
> If I understand the implementation correctly, even non-animated string properties will be converted to a double.

Nevermind. I see it now. Only converted to double if animated. I think this is our best tradeoff - unless we want to implement number detection.

Reply:
In hindsight, per my comment above about developers and generic UIs, I think now we should not set `animation: yes` (a new attribute in schema 7.0) in any of the filter metadata to prevent that. Basically, a user or developer will have to do their own analysis and testing to deliberately opt into using an animated property value with an avfilter.

Reply:
I agree with that. Maybe just add a note in the metadata header as a breadcrumb for future developers (or ourselves).